io.camunda.zeebe.engine.processing.authorization.RoleTest causing JVM crashes

**Describe the bug**

I'm opening a bug issue instead of a flaky test, because the consequence is not just that the test is flaky, but that it's crashing the JVM. I would recommend we consider this a bug to start with, and downgrade it to flaky test if we identify it's _only_ occurring in tests.

Happened in this run: https://github.com/camunda/camunda/actions/runs/11502526769/job/32017737075

**To Reproduce**

Unclear yet, but I suppose re-running the test would work.

**Expected behavior**

JVM does not crash :)

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
ExecutionException The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
Error:  Command was /bin/sh -c cd '/home/runner/_work/camunda/camunda/zeebe/engine' && '/home/runner/_work/_tool/Java_Temurin-Hotspot_jdk/21.0.5-11.0.LTS/x64/bin/java' '-javaagent:/home/runner/.m2/repository/cached/releases/org/jacoco/org.jacoco.agent/0.8.12/org.jacoco.agent-0.8.12-runtime.jar=destfile=/home/runner/_work/camunda/camunda/zeebe/engine/target/jacoco.exec' '-jar' '/home/runner/_work/camunda/camunda/zeebe/engine/target/surefire/surefirebooter-20241024153602999_1725.jar' '/home/runner/_work/camunda/camunda/zeebe/engine/target/surefire' '2024-10-24T15-31-42_949-jvmRun3' 'surefire-20241024153602999_1723tmp' 'surefire_214-20241024153602999_1724tmp'
Error:  Error occurred in starting fork, check output in log
Error:  Process Exit Code: 134
Error:  Crashed tests:
Error:  io.camunda.zeebe.engine.processing.authorization.RoleTest
Error:  org.apache.maven.surefire.booter.SurefireBooterForkException: ExecutionException The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
Error:  Command was /bin/sh -c cd '/home/runner/_work/camunda/camunda/zeebe/engine' && '/home/runner/_work/_tool/Java_Temurin-Hotspot_jdk/21.0.5-11.0.LTS/x64/bin/java' '-javaagent:/home/runner/.m2/repository/cached/releases/org/jacoco/org.jacoco.agent/0.8.12/org.jacoco.agent-0.8.12-runtime.jar=destfile=/home/runner/_work/camunda/camunda/zeebe/engine/target/jacoco.exec' '-jar' '/home/runner/_work/camunda/camunda/zeebe/engine/target/surefire/surefirebooter-20241024153602999_1725.jar' '/home/runner/_work/camunda/camunda/zeebe/engine/target/surefire' '2024-10-24T15-31-42_949-jvmRun3' 'surefire-20241024153602999_1723tmp' 'surefire_214-20241024153602999_1724tmp'
Error:  Error occurred in starting fork, check output in log
Error:  Process Exit Code: 134
Error:  Crashed tests:
Error:  io.camunda.zeebe.engine.processing.authorization.RoleTest
Error:  	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.awaitResultsDone(ForkStarter.java:456)
Error:  	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.runSuitesForkOnceMultiple(ForkStarter.java:358)
Error:  	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:296)
Error:  	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:250)
Error:  	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1336)
Error:  	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1134)
Error:  	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:968)
Error:  	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:126)
Error:  	at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2(MojoExecutor.java:328)
Error:  	at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute(MojoExecutor.java:316)
Error:  	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
Error:  	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:174)
Error:  	at org.apache.maven.lifecycle.internal.MojoExecutor.access$000(MojoExecutor.java:75)
Error:  	at org.apache.maven.lifecycle.internal.MojoExecutor$1.run(MojoExecutor.java:162)
Error:  	at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute(DefaultMojosExecutionStrategy.java:39)
Error:  	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:159)
Error:  	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:105)
Error:  	at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:193)
Error:  	at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:180)
Error:  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
Error:  	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
Error:  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
Error:  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
Error:  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
Error:  	at java.base/java.lang.Thread.run(Thread.java:1583)
Error:  Caused by: org.apache.maven.surefire.booter.SurefireBooterForkException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
Error:  Command was /bin/sh -c cd '/home/runner/_work/camunda/camunda/zeebe/engine' && '/home/runner/_work/_tool/Java_Temurin-Hotspot_jdk/21.0.5-11.0.LTS/x64/bin/java' '-javaagent:/home/runner/.m2/repository/cached/releases/org/jacoco/org.jacoco.agent/0.8.12/org.jacoco.agent-0.8.12-runtime.jar=destfile=/home/runner/_work/camunda/camunda/zeebe/engine/target/jacoco.exec' '-jar' '/home/runner/_work/camunda/camunda/zeebe/engine/target/surefire/surefirebooter-20241024153602999_1725.jar' '/home/runner/_work/camunda/camunda/zeebe/engine/target/surefire' '2024-10-24T15-31-42_949-jvmRun3' 'surefire-20241024153602999_1723tmp' 'surefire_214-20241024153602999_1724tmp'
Error:  Error occurred in starting fork, check output in log
Error:  Process Exit Code: 134
Error:  Crashed tests:
Error:  io.camunda.zeebe.engine.processing.authorization.RoleTest
Error:  	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:643)
Error:  	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.lambda$null$3(ForkStarter.java:350)
Error:  	... 4 more
```

</p>
</details>

--- a/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/authorization/RoleTest.java
+++ b/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/authorization/RoleTest.java
@@ -8,11 +8,7 @@
 package io.camunda.zeebe.engine.processing.authorization;
 
 import static io.camunda.zeebe.protocol.record.Assertions.assertThat;
-import static org.junit.Assert.assertTrue;
 
-import io.camunda.zeebe.engine.state.immutable.AuthorizationState;
-import io.camunda.zeebe.engine.state.immutable.RoleState;
-import io.camunda.zeebe.engine.state.immutable.UserState;
 import io.camunda.zeebe.engine.util.EngineRule;
 import io.camunda.zeebe.protocol.record.RejectionType;
 import io.camunda.zeebe.protocol.record.value.AuthorizationOwnerType;
@@ -22,54 +18,32 @@ import io.camunda.zeebe.protocol.record.value.PermissionAction;
 import io.camunda.zeebe.test.util.record.RecordingExporterTestWatcher;
 import java.util.UUID;
 import org.assertj.core.api.Assertions;
-import org.junit.BeforeClass;
-import org.junit.ClassRule;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TestWatcher;
 
 public class RoleTest {
 
-  @ClassRule public static final EngineRule ENGINE = EngineRule.singlePartition();
-  private static long userKey;
+  @Rule public final EngineRule engine = EngineRule.singlePartition();
   @Rule public final TestWatcher recordingExporterTestWatcher = new RecordingExporterTestWatcher();
-  private final RoleState roleState = ENGINE.getProcessingState().getRoleState();
-  private final UserState userState = ENGINE.getProcessingState().getUserState();
-  private final AuthorizationState authorizationState =
-      ENGINE.getProcessingState().getAuthorizationState();
-
-  @BeforeClass
-  public static void setUp() {
-    userKey =
-        ENGINE
-            .user()
-            .newUser("foo")
-            .withEmail("foo@bar")
-            .withName("Foo Bar")
-            .withPassword("zabraboof")
-            .create()
-            .getKey();
-  }
 
   @Test
   public void shouldCreateRole() {
     final var name = UUID.randomUUID().toString();
-    final var roleRecord = ENGINE.role().newRole(name).create();
+    final var roleRecord = engine.role().newRole(name).create();
 
     final var createdRole = roleRecord.getValue();
     Assertions.assertThat(createdRole).isNotNull().hasFieldOrPropertyWithValue("name", name);
-    final var ownerType = authorizationState.getOwnerType(createdRole.getRoleKey());
-    Assertions.assertThat(ownerType).isPresent().hasValue(AuthorizationOwnerType.ROLE);
   }
 
   @Test
   public void shouldNotDuplicate() {
     // given
     final var name = UUID.randomUUID().toString();
-    final var roleRecord = ENGINE.role().newRole(name).create();
+    final var roleRecord = engine.role().newRole(name).create();
 
     // when
-    final var duplicatedRoleRecord = ENGINE.role().newRole(name).expectRejection().create();
+    final var duplicatedRoleRecord = engine.role().newRole(name).expectRejection().create();
 
     final var createdRole = roleRecord.getValue();
     Assertions.assertThat(createdRole).isNotNull().hasFieldOrPropertyWithValue("name", name);
@@ -86,12 +60,12 @@ public class RoleTest {
   public void shouldUpdateRole() {
     // given
     final var name = UUID.randomUUID().toString();
-    final var createdRecord = ENGINE.role().newRole(name).create();
+    final var createdRecord = engine.role().newRole(name).create();
 
     // when
     final var newName = UUID.randomUUID().toString();
     final var updatedRoleRecord =
-        ENGINE.role().updateRole(createdRecord.getValue().getRoleKey()).withName(newName).update();
+        engine.role().updateRole(createdRecord.getValue().getRoleKey()).withName(newName).update();
 
     final var updatedRole = updatedRoleRecord.getValue();
     Assertions.assertThat(updatedRole).isNotNull().hasFieldOrPropertyWithValue("name", newName);
@@ -101,12 +75,12 @@ public class RoleTest {
   public void shouldRejectIfRoleIsNotPresent() {
     // given
     final var name = UUID.randomUUID().toString();
-    final var roleRecord = ENGINE.role().newRole(name).create();
+    final var roleRecord = engine.role().newRole(name).create();
 
     // when
     final var notPresentRoleKey = 1L;
     final var notPresentUpdateRecord =
-        ENGINE.role().updateRole(notPresentRoleKey).expectRejection().update();
+        engine.role().updateRole(notPresentRoleKey).expectRejection().update();
 
     final var createdRole = roleRecord.getValue();
     Assertions.assertThat(createdRole).isNotNull().hasFieldOrPropertyWithValue("name", name);
@@ -123,13 +97,13 @@ public class RoleTest {
   public void shouldRejectIfRoleWithSameNameIsPresent() {
     // given
     final var name = UUID.randomUUID().toString();
-    final var roleKey = ENGINE.role().newRole(name).create().getValue().getRoleKey();
+    final var roleKey = engine.role().newRole(name).create().getValue().getRoleKey();
     final var anotherName = UUID.randomUUID().toString();
-    ENGINE.role().newRole(anotherName).create();
+    engine.role().newRole(anotherName).create();
 
     // when
     final var notPresentUpdateRecord =
-        ENGINE.role().updateRole(roleKey).withName(anotherName).expectRejection().update();
+        engine.role().updateRole(roleKey).withName(anotherName).expectRejection().update();
 
     assertThat(notPresentUpdateRecord)
         .hasRejectionType(RejectionType.ALREADY_EXISTS)
@@ -141,10 +115,19 @@ public class RoleTest {
 
   @Test
   public void shouldAddEntityToRole() {
+    final var userKey =
+        engine
+            .user()
+            .newUser("foo")
+            .withEmail("foo@bar")
+            .withName("Foo Bar")
+            .withPassword("zabraboof")
+            .create()
+            .getKey();
     final var name = UUID.randomUUID().toString();
-    final var roleKey = ENGINE.role().newRole(name).create().getValue().getRoleKey();
+    final var roleKey = engine.role().newRole(name).create().getValue().getRoleKey();
     final var updatedRole =
-        ENGINE
+        engine
             .role()
             .addEntity(roleKey)
             .withEntityKey(userKey)
@@ -154,26 +137,20 @@ public class RoleTest {
 
     Assertions.assertThat(updatedRole)
         .isNotNull()
-        .hasFieldOrPropertyWithValue("entityKey", userKey);
-
-    final var persistedEntity = roleState.getEntityType(roleKey, userKey);
-    assertTrue(persistedEntity.isPresent());
-    Assertions.assertThat(persistedEntity.get()).isEqualTo(EntityType.USER);
-    final var user = userState.getUser(userKey);
-    assertTrue(user.isPresent());
-    Assertions.assertThat(user.get().getRoleKeysList()).contains(roleKey);
+        .hasFieldOrPropertyWithValue("entityKey", userKey)
+        .hasFieldOrPropertyWithValue("entityType", EntityType.USER);
   }
 
   @Test
   public void shouldRejectIfRoleIsNotPresentWhileAddingEntity() {
     // given
     final var name = UUID.randomUUID().toString();
-    final var roleRecord = ENGINE.role().newRole(name).create();
+    final var roleRecord = engine.role().newRole(name).create();
 
     // when
     final var notPresentRoleKey = 1L;
     final var notPresentUpdateRecord =
-        ENGINE.role().addEntity(notPresentRoleKey).expectRejection().add();
+        engine.role().addEntity(notPresentRoleKey).expectRejection().add();
 
     final var createdRole = roleRecord.getValue();
     Assertions.assertThat(createdRole).isNotNull().hasFieldOrPropertyWithValue("name", name);
@@ -190,13 +167,13 @@ public class RoleTest {
   public void shouldRejectIfEntityIsNotPresent() {
     // given
     final var name = UUID.randomUUID().toString();
-    final var roleRecord = ENGINE.role().newRole(name).create();
+    final var roleRecord = engine.role().newRole(name).create();
 
     // when
     final var createdRole = roleRecord.getValue();
     final var roleKey = createdRole.getRoleKey();
     final var notPresentUpdateRecord =
-        ENGINE
+        engine
             .role()
             .addEntity(roleKey)
             .withEntityKey(1L)
@@ -215,34 +192,44 @@ public class RoleTest {
 
   @Test
   public void shouldRemoveEntityToRole() {
+    final var userKey =
+        engine
+            .user()
+            .newUser("foo")
+            .withEmail("foo@bar")
+            .withName("Foo Bar")
+            .withPassword("zabraboof")
+            .create()
+            .getKey();
     final var name = UUID.randomUUID().toString();
-    final var roleKey = ENGINE.role().newRole(name).create().getValue().getRoleKey();
-    ENGINE.role().addEntity(roleKey).withEntityKey(userKey).withEntityType(EntityType.USER).add();
-    ENGINE
-        .role()
-        .removeEntity(roleKey)
-        .withEntityKey(userKey)
-        .withEntityType(EntityType.USER)
-        .remove()
-        .getValue();
+    final var roleKey = engine.role().newRole(name).create().getValue().getRoleKey();
+    engine.role().addEntity(roleKey).withEntityKey(userKey).withEntityType(EntityType.USER).add();
+    final var removedEntity =
+        engine
+            .role()
+            .removeEntity(roleKey)
+            .withEntityKey(userKey)
+            .withEntityType(EntityType.USER)
+            .remove()
+            .getValue();
 
-    final var persistedEntity = roleState.getEntityType(roleKey, userKey);
-    assertTrue(persistedEntity.isEmpty());
-    final var user = userState.getUser(userKey);
-    assertTrue(user.isPresent());
-    Assertions.assertThat(user.get().getRoleKeysList()).isEmpty();
+    Assertions.assertThat(removedEntity)
+        .isNotNull()
+        .hasFieldOrPropertyWithValue("roleKey", roleKey)
+        .hasFieldOrPropertyWithValue("entityKey", userKey)
+        .hasFieldOrPropertyWithValue("entityType", EntityType.USER);
   }
 
   @Test
   public void shouldRejectIfRoleIsNotPresentEntityRemoval() {
     // given
     final var name = UUID.randomUUID().toString();
-    final var roleRecord = ENGINE.role().newRole(name).create();
+    final var roleRecord = engine.role().newRole(name).create();
 
     // when
     final var notPresentRoleKey = 1L;
     final var notPresentUpdateRecord =
-        ENGINE.role().addEntity(notPresentRoleKey).expectRejection().add();
+        engine.role().addEntity(notPresentRoleKey).expectRejection().add();
 
     final var createdRole = roleRecord.getValue();
     Assertions.assertThat(createdRole).isNotNull().hasFieldOrPropertyWithValue("name", name);
@@ -259,13 +246,13 @@ public class RoleTest {
   public void shouldRejectIfEntityIsNotPresentEntityRemoval() {
     // given
     final var name = UUID.randomUUID().toString();
-    final var roleRecord = ENGINE.role().newRole(name).create();
+    final var roleRecord = engine.role().newRole(name).create();
 
     // when
     final var createdRole = roleRecord.getValue();
     final var roleKey = createdRole.getRoleKey();
     final var notPresentUpdateRecord =
-        ENGINE
+        engine
             .role()
             .removeEntity(roleKey)
             .withEntityKey(1L)
@@ -286,8 +273,8 @@ public class RoleTest {
   public void shouldDeleteRole() {
     // given
     final var name = UUID.randomUUID().toString();
-    final var roleKey = ENGINE.role().newRole(name).create().getValue().getRoleKey();
-    ENGINE
+    final var roleKey = engine.role().newRole(name).create().getValue().getRoleKey();
+    engine
         .authorization()
         .permission()
         .withOwnerKey(roleKey)
@@ -297,10 +284,9 @@ public class RoleTest {
         .add();
 
     // when
-    ENGINE.role().deleteRole(roleKey).delete();
+    final var deletedRole = engine.role().deleteRole(roleKey).delete().getValue();
 
-    final var deletedRecord = roleState.getRole(roleKey);
-    assertTrue(deletedRecord.isEmpty());
+    Assertions.assertThat(deletedRole).isNotNull().hasFieldOrPropertyWithValue("roleKey", roleKey);
   }
 
   @Test
@@ -308,7 +294,7 @@ public class RoleTest {
     // when
     final var notPresentRoleKey = 1L;
     final var notPresentUpdateRecord =
-        ENGINE.role().deleteRole(notPresentRoleKey).expectRejection().delete();
+        engine.role().deleteRole(notPresentRoleKey).expectRejection().delete();
 
     assertThat(notPresentUpdateRecord)
         .hasRejectionType(RejectionType.NOT_FOUND)

--- a/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/tenant/AddEntityTenantMultiPartitionTest.java
+++ b/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/tenant/AddEntityTenantMultiPartitionTest.java
@@ -26,21 +26,20 @@ import java.time.Duration;
 import java.util.UUID;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.stream.Collectors;
-import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TestWatcher;
 
 public class AddEntityTenantMultiPartitionTest {
   private static final int PARTITION_COUNT = 3;
-  private static long userKey;
 
   @Rule public final EngineRule engine = EngineRule.multiplePartition(PARTITION_COUNT);
   @Rule public final TestWatcher testWatcher = new RecordingExporterTestWatcher();
 
-  @Before
-  public void setUp() {
-    userKey =
+  @Test
+  public void shouldDistributeTenantAddEntityCommand() {
+    // when
+    final var userKey =
         engine
             .user()
             .newUser("foo")
@@ -49,11 +48,6 @@ public class AddEntityTenantMultiPartitionTest {
             .withPassword("zabraboof")
             .create()
             .getKey();
-  }
-
-  @Test
-  public void shouldDistributeTenantAddEntityCommand() {
-    // when
     final var tenantId = UUID.randomUUID().toString();
     final var tenantKey =
         engine.tenant().newTenant().withTenantId(tenantId).create().getValue().getTenantKey();
@@ -106,6 +100,15 @@ public class AddEntityTenantMultiPartitionTest {
   @Test
   public void shouldDistributeInIdentityQueue() {
     // when
+    final var userKey =
+        engine
+            .user()
+            .newUser("foo")
+            .withEmail("foo@bar")
+            .withName("Foo Bar")
+            .withPassword("zabraboof")
+            .create()
+            .getKey();
     final var tenantId = UUID.randomUUID().toString();
     final var tenantKey =
         engine.tenant().newTenant().withTenantId(tenantId).create().getValue().getTenantKey();
@@ -128,6 +131,16 @@ public class AddEntityTenantMultiPartitionTest {
   @Test
   public void distributionShouldNotOvertakeOtherCommandsInSameQueue() {
     // given the tenant creation distribution is intercepted
+    final var userKey =
+        engine
+            .user()
+            .newUser("foo")
+            .withEmail("foo@bar")
+            .withName("Foo Bar")
+            .withPassword("zabraboof")
+            .create()
+            .getKey();
+
     for (int partitionId = 2; partitionId <= PARTITION_COUNT; partitionId++) {
       interceptTenantCreateForPartition(partitionId); // Intercept tenant creation
     }

--- a/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/tenant/UpdateTenantMultiPartitionTest.java
+++ b/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/tenant/UpdateTenantMultiPartitionTest.java
@@ -63,7 +63,8 @@ public class UpdateTenantMultiPartitionTest {
     assertThat(
             RecordingExporter.records()
                 .withPartitionId(1)
-                .limit(record -> record.getIntent().equals(CommandDistributionIntent.FINISHED)))
+                .limitByCount(
+                    record -> record.getIntent().equals(CommandDistributionIntent.FINISHED), 2))
         .extracting(
             Record::getIntent,
             Record::getRecordType,
@@ -124,12 +125,12 @@ public class UpdateTenantMultiPartitionTest {
   @Test
   public void distributionShouldNotOvertakeOtherCommandsInSameQueue() {
     // when
-    final var tenantId = UUID.randomUUID().toString();
-    final var tenantKey =
-        engine.tenant().newTenant().withTenantId(tenantId).create().getValue().getTenantKey();
     for (int partitionId = 2; partitionId <= PARTITION_COUNT; partitionId++) {
       interceptTenantCreateForPartition(partitionId);
     }
+    final var tenantId = UUID.randomUUID().toString();
+    final var tenantKey =
+        engine.tenant().newTenant().withTenantId(tenantId).create().getValue().getTenantKey();
     engine.tenant().updateTenant(tenantKey).withTenantId(tenantId + "-updated").update();
 
     // Increase time to trigger a redistribution

